libdevice/sanitizer_utils.cpp

```diff
@@ -6,22 +6,627 @@
+    __DeviceSanitizerReportMem.LID1 = __spirv_LocalInvocationId_y();
+    __DeviceSanitizerReportMem.LID2 = __spirv_LocalInvocationId_z();
+
+    __DeviceSanitizerReportMem.IsWrite = is_write;
+    __DeviceSanitizerReportMem.AccessSize = access_size;
+    __DeviceSanitizerReportMem.ErrorType = error_type;
+    __DeviceSanitizerReportMem.MemoryType = memory_type;
+    __DeviceSanitizerReportMem.IsRecover = is_recover;
+
+    // Show we've done copying
+    atomicStore(&__DeviceSanitizerReportMem.Flag, ASAN_REPORT_FINISH);
+  }
+}
+
+///
+/// ASAN Error Reporters
+///
+
+void __asan_report_overlapping_error(uptr dst, uptr src, size_t size,
+                                     const char *file, int32_t line,
+                                     const char *func) {
+  // TODO
+}
+
+void __asan_report_shadow_error(uptr addr, int32_t as, size_t size,
+                                bool is_write, uptr poisoned_addr,
+                                const char *file, int32_t line,
+                                const char *func) {
+  // TODO
+}
+
+void __asan_report_access_error(uptr addr, int32_t as, size_t size,
+                                bool is_write, uptr poisoned_addr,
+                                const char __SYCL_CONSTANT__ *file,
+                                int32_t line,
+                                const char __SYCL_CONSTANT__ *func,
+                                bool is_recover = false) {
+  // Check Error Type
+  s8 *shadow_address = (s8 *)MemToShadow(poisoned_addr, as);
+  int shadow_value = *shadow_address;
+  if (shadow_value > 0) {
+    shadow_value = *(shadow_address + 1);
+  }
+  // FIXME: check if shadow_address out-of-bound
+
+  // print_shadow_memory(addr);
+
+  // __spirv_ocl_printf(__asan_current_shadow_value, shadow_value);
+
+  DeviceSanitizerMemoryType memory_type;
+  DeviceSanitizerErrorType error_type;
+
+  switch (shadow_value) {
+  case kUsmDeviceRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_DEVICE;
+    error_type = DeviceSanitizerErrorType::OUT_OF_BOUND;
+    break;
+  case kUsmHostRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_HOST;
+    error_type = DeviceSanitizerErrorType::OUT_OF_BOUND;
+    break;
+  case kUsmSharedRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_SHARED;
+    error_type = DeviceSanitizerErrorType::OUT_OF_BOUND;
+    break;
+  case kUsmDeviceDeallocatedMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_DEVICE;
+    error_type = DeviceSanitizerErrorType::USE_AFTER_FREE;
+    break;
+  case kUsmHostDeallocatedMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_HOST;
+    error_type = DeviceSanitizerErrorType::USE_AFTER_FREE;
+    break;
+  case kUsmSharedDeallocatedMagic:
+    memory_type = DeviceSanitizerMemoryType::USM_SHARED;
+    error_type = DeviceSanitizerErrorType::USE_AFTER_FREE;
+    break;
+  case kPrivateLeftRedzoneMagic:
+    memory_type = DeviceSanitizerMemoryType::PRIVATE;
```

Comment:
Can these 3"case" be merged because the setting for memory_type and error_type are same?

Reply:
Nice catch! Done